Common/config/ruledtaskbarappearance.hpp

```diff
@@ -0,0 +1,122 @@
+#pragma once
+#include <string_view>
+#include <vector>
+#include <format>
+
+#include "rapidjsonhelper.hpp"
+#include "optionaltaskbarappearance.hpp"
+#include "rule.hpp"
+#include "../../TranslucentTB/windows/window.hpp"
+
+struct RuledTaskbarAppearance : OptionalTaskbarAppearance {
+	std::vector<Rule> Rules = {};
+
+	constexpr RuledTaskbarAppearance() noexcept = default;
+	constexpr RuledTaskbarAppearance(bool enabled, ACCENT_STATE accent, Util::Color color, bool showPeek, std::vector<Rule> rules) noexcept :
+		OptionalTaskbarAppearance(enabled, accent, color, showPeek),
+		Rules(rules)
+	{ }
+
+	template<typename Writer>
+	inline void Serialize(Writer& writer) const
+	{
+		OptionalTaskbarAppearance::Serialize(writer);
+		SerializeRulesVec(writer, Rules, RULES_KEY);
+	}
+
+	void Deserialize(const rjh::value_t& obj, void (*unknownKeyCallback)(std::wstring_view))
+	{
+		rjh::EnsureType(rj::Type::kObjectType, obj.GetType(), L"root node");
+
+		for (auto it = obj.MemberBegin(); it != obj.MemberEnd(); ++it)
+		{
+			rjh::EnsureType(rj::Type::kStringType, it->name.GetType(), L"member name");
+
+			const auto key = rjh::ValueToStringView(it->name);
+			if (key == RULES_KEY)
+			{
+				DeserializeRulesVec(it->value, Rules, key, unknownKeyCallback);
+			}
+			else
+			{
+				OptionalInnerDeserialize(key, it->value, unknownKeyCallback);
+			}
+		}
+	}
+
+	inline const std::optional<Rule> FindRule(const Window* window) const
+	{
+		const auto windowClassName = window->classname();
+		const auto windowFile = window->file();
+		const auto windowFileName = windowFile->filename().native();
+		const auto windowTitle = window->title();
+
+		for (const Rule rule : Rules)
+		{
+			// This is the fastest because we do the less string manipulation, so always try it first
+			if (!rule.m_Class.empty())
+			{
+				if (windowClassName)
```

Comment:
You can just write `rule.m_Class == windowClassName`, std::optional has a built-in operator== that returns false if it's not holding anything